scalameta/semanticdb-scalac/src/main/scala/scala/meta/internal/semanticdb/ConfigOps.scala

```diff
@@ -5,48 +5,129 @@ import scala.meta.internal.io.PathIO
+      s"-P:${SemanticdbPlugin.name}:synthetics:${synthetics.name} "
 }
 object SemanticdbConfig {
-  def default = SemanticdbConfig(PathIO.workingDirectory, SemanticdbMode.Fat, FailureMode.Warning)
+  def default =
+    SemanticdbConfig(
+      PathIO.workingDirectory,
+      SemanticdbMode.Fat,
+      FailureMode.Warning,
+      DenotationMode.All,
+      ProfilingMode.Off,
+      ".*",
+      "",
+      MessageMode.All,
+      SyntheticMode.All)
+}
+
+sealed abstract class SemanticdbMode {
+  def name: String = toString.toLowerCase
+  import SemanticdbMode._
+  def isSlim: Boolean = this == Slim
+  def isFat: Boolean = this == Fat
+  def isDisabled: Boolean = this == Disabled
+}
+object SemanticdbMode {
+  def unapply(arg: String): Option[SemanticdbMode] = all.find(_.toString.equalsIgnoreCase(arg))
+  def all = List(Fat, Slim, Disabled)
+  case object Fat extends SemanticdbMode
+  case object Slim extends SemanticdbMode
+  case object Disabled extends SemanticdbMode
 }
 
 sealed abstract class FailureMode {
   def name: String = toString.toLowerCase
 }
 object FailureMode {
-  def unapply(arg: String): Option[FailureMode] =
-    all.find(_.toString.equalsIgnoreCase(arg))
+  def unapply(arg: String): Option[FailureMode] = all.find(_.toString.equalsIgnoreCase(arg))
   def all = List(Error, Warning, Info, Ignore)
   case object Error extends FailureMode
   case object Warning extends FailureMode
   case object Info extends FailureMode
   case object Ignore extends FailureMode
 }
 
-sealed abstract class SemanticdbMode {
-  import SemanticdbMode._
-  def isSlim: Boolean = this == Slim
-  def isFat: Boolean = this == Fat
-  def isDisabled: Boolean = this == Disabled
+sealed abstract class DenotationMode {
+  def name: String = toString.toLowerCase
+  import DenotationMode._
+  def saveDefinitions: Boolean = this == All || this == Definitions
+  def saveReferences: Boolean = this == All
 }
-object SemanticdbMode {
-  def unapply(arg: String): Option[SemanticdbMode] =
-    all.find(_.toString.equalsIgnoreCase(arg))
-  def all = List(Fat, Slim, Disabled)
-  case object Fat extends SemanticdbMode
-  case object Slim extends SemanticdbMode
-  case object Disabled extends SemanticdbMode
+object DenotationMode {
+  def unapply(arg: String): Option[DenotationMode] = all.find(_.toString.equalsIgnoreCase(arg))
+  def all = List(All, Definitions, None)
+  case object All extends DenotationMode
+  case object Definitions extends DenotationMode
+  case object None extends DenotationMode
+}
+
+sealed abstract class ProfilingMode {
+  def name: String = toString.toLowerCase
+  import ProfilingMode._
+  def isConsole: Boolean = this == Console
+  def isOff: Boolean = this == Off
+}
+object ProfilingMode {
+  def unapply(arg: String): Option[ProfilingMode] = all.find(_.toString.equalsIgnoreCase(arg))
+  def all = List(Console, Off)
+  case object Console extends ProfilingMode
+  case object Off extends ProfilingMode
+}
+
+sealed abstract class MessageMode {
```

Comment:
Related https://github.com/scalameta/scalameta/issues/1151 I think we could actually avoid a ton of boilerplate here if we define these enums in semanticdb.proto and persist them in each Document or Database. We don't have to expose them in `scala.meta._`, they would only exists in the schema layer.